registration/include/pcl/registration/correspondence_estimation_lookup_table.h

```diff
@@ -0,0 +1,517 @@
+         * @param[out] correspondence_index The computed index of the lookup table cell.
+         * @param[out] correspondence_index_components The computed index [x,y,z] components.
+         * @return True if the index is valid. False if the query point was outside of the lookup table bounds.
+         */
+        bool
+        computeCorrespondenceCellIndex (const PointT& query_point, size_t& correspondence_index, Eigen::Vector3i& correspondence_index_components);
+
+        /**
+         * Computes the correspondence cell centroid from the lookup table cell index components (retrieved with @see computeCorrespondenceCellIndex).
+         * @return The centroid of the correspondence cell.
+         */
+        Eigen::Vector3f
+        computeCorrespondenceCellCentroid (const Eigen::Vector3i& correspondence_index_components);
+
+        /**
+         * Computes the squared distance between two points.
+         * @param a First point.
+         * @param b Second point.
+         * @return The Euclidean squared distance between the two points.
+         */
+        float
+        computeSquaredDistance (const Eigen::Vector3f& a, const Eigen::Vector3f& b)
+        {
+          return ((a - b).squaredNorm ());
+        }
+
+        /**
+         * Computes the squared distance between two points.
+         * @param a First point.
+         * @param b Second point.
+         * @return The Euclidean squared distance between the two points.
+         */
+        float
+        computeSquaredDistance (const PointT& a, const PointT& b)
+        {
+          float dx = a.x - b.x;
+          float dy = a.y - b.y;
+          float dz = a.z - b.z;
+          return (dx * dx + dy * dy + dz * dz);
+        }
+
+        /**
+         * Computes the squared distance between the query point and the cell centroid.
+         * @param query_point Query point.
+         * @param centroid Cell centroid.
+         * @return The Euclidean squared distance between the two points.
+         */
+        float
+        computeSquaredDistanceToCentroid (const PointT& query_point, Eigen::Vector3f centroid)
+        {
+          float dx = query_point.x - centroid (0);
+          float dy = query_point.y - centroid (1);
+          float dz = query_point.z - centroid (2);
+          return (dx * dx + dy * dy + dz * dz);
+        }
+
+        /**
+         * Retrieves the point associated to a given cell by using the index of the point cloud that was used to initialize the lookup table.
+         * @param[in] index Index of the point to retrieve.
+         * @param[out] point Point retrieved.
+         * @return True if the index was valid and the point was retrieved.
+         */
+        bool
+        getCellPoint (size_t index, PointT& point)
+        {
+          if (pointcloud_ && index < pointcloud_->size ())
+          {
+            point = (*pointcloud_)[index];
+            return (true);
+          }
+          return (false);
+        }
```

Comment:
Is this really necessary? It's a method to do bound checks on a vector... This can be checked in advance if really needed. Most of the times don't always need to check for the validity of the index.